br/pkg/encryption/master_key/multi_master_key_backend.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 PingCAP, Inc. Licensed under Apache-2.0.
+
+package encryption
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/encryptionpb"
+)
+
+const (
+	defaultBackendCapacity = 5
+)
+
+type MultiMasterKeyBackend struct {
+	backends []Backend
+}
+
+func NewMultiMasterKeyBackend(masterKeysProto []*encryptionpb.MasterKey) (*MultiMasterKeyBackend, error) {
+	if masterKeysProto == nil && len(masterKeysProto) == 0 {
+		return nil, errors.New("must provide at least one master key")
+	}
+	var backends = make([]Backend, 0, defaultBackendCapacity)
+	for _, masterKeyProto := range masterKeysProto {
+		backend, err := CreateBackend(masterKeyProto)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		backends = append(backends, backend)
+	}
+	return &MultiMasterKeyBackend{
+		backends: backends,
+	}, nil
+}
+
+func (m *MultiMasterKeyBackend) Decrypt(ctx context.Context, encryptedContent *encryptionpb.EncryptedContent) ([]byte, error) {
```

Comment:
Perhaps we don't need to use pointer receiver type, because this neither modifiy the `Backend` itself nor the `Backend` value is too huge to be passed by value effectively.

Reply:
hmm, im not a Go expert so probably wrong. If we return a copy, the original one will be immediately subject to GC, which adds the GC pressure. Returning pointer feels cheaper to me, and I couldn't think of the downside of returning pointer compared to a copy. Most likely I'm missing the benefit of returning a copy, let me know your thoughts!

Reply:
Because slice is in fact a value type with 3 fields: `cap`, `len` and `data`(pointer to the real content). In fact it is trivial to construct and destruct such a object, it will be kept at stack, GC will probably not be involved. 

When we are using pointers, we may need extra indirection to access the content. Also a pointer object may leak to heap, which needs GC to clean up. 

Anyway this is a nit, don't mind. You may get more information about go slices here: https://go.dev/blog/slices-intro.

Reply:
Got it, thanks for the info! yeah I get it that slice in this case is trivial to copy, I'm not certain about the practice of returning the copy of a struct or the pointer of it, I guess I need to check the GO memory model, cuz in Java every instance of a class is in memory. 

Do you have any recommendations of when to return pointer and when to return a copy? I understand when a struct is heavy we should return a pointer but I'm not sure what's the drawing line between heavy and not heavy. Thanks